imageio/core/imopen.py

```diff
@@ -139,14 +139,26 @@ def imopen(
     # plugin specified, no search needed
     # (except in legacy mode)
     if plugin is not None:
-        try:
-            config = _get_config(plugin, legacy_mode)
-        except (IndexError, ValueError):
-            request.finish()
-            raise
+        if isinstance(plugin, str):
+            try:
+                config = _get_config(plugin, legacy_mode)
+            except (IndexError, ValueError):
+                request.finish()
+                raise
+
+            def loader(request, **kwargs):
+                return config.plugin_class(request, **kwargs)
+
+        elif not legacy_mode:
```

Comment:
I know we cant do isinstance here, but maybe do a `hasattr` to see whether this looks like a plugin?

Reply:
That would be a pretty extensive test, but it is certainly possible. Currently, we would have to run that check for `read`, `write`, `iter`, `close`, `get_meta`, `__enter__` and `_exit__`.  We could do this cleanly by factoring this out into a separate `looks_like_a_plugin` function. However, at this point why don't we ask people to inherit from a common abstract `Plugin`? I am currently writing one anyway as a means of documenting the plugin interface. people will already have to depend on ImageIO to get access to the `Request` object used by `Plugin.__init__(request:Request, **kwargs)` so it won't be a big change.

Also, do you think it will happen often that we encounter classes/objects that (accidentally) succeed on `MyClass(request, **kwargs)`, but lack the plugin interface? It seems like a quite specific thing to "accidentally get right", and the cases that do fail we catch in the subsequent `try` block already. I guess if you try hard enough then this isn't bulletproof validation, but in that case `hasattr` won't stand much ground either.

My current thinking was to do it "numpy style" and omit explicit input validation (beyond what is needed to decide control flow). If some functionality that we didn't document works, good for you. If it breaks (potentially with a cryptic error message) then because you are doing non-standard stuff. Not sure to what extent we should try to catch these cases and try to provide clear error messages.

Reply:
I was thinking more along the lines of checking one or two attributes, mainly to avoid the use case of accidentally providing something that is definitely not a plugin.

If a non-plugin is provided now, how will it fail, with what error message. If this message is clear enough to see what went wrong, then lets not bother. But if its hard to deduce that this is the point where it went wrong, then it would be worth adding a check here.

Reply:
Assiming the constructor fails (e.g., because it doesn't know what to do with `Request` and ends up doing the wrong thing), then this piece of code will catch that:

https://github.com/imageio/imageio/blob/924e1c5c43576acf16247c3c47b7baf6c6f68fe6/imageio/core/imopen.py#L161-L164

so the error looks something like this:

```plugin
>>> img = iio.v3.imread("imageio:chelsea.png", plugin=dict)         
Traceback (most recent call last):
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\imopen.py", line 161, in imopen
    return loader(request, **kwargs)
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\imopen.py", line 155, in loader
    return plugin(request, **kwargs)
TypeError: 'Request' object is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\v3_api.py", line 39, in imread
    with imopen(uri, "r", **plugin_kwargs) as img_file:
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\imopen.py", line 179, in imopen
    raise err_type(err_msg) from err_from
OSError: An unknown error occured while initializing `<class 'dict'>`
```

Reply:
Right, perhaps instead of doing a check, we can catch those errors and reraise with a better error message? If you don't think this makes sense, then let's go with what we have now :)

Reply:
> Right, perhaps instead of doing a check, we can catch those errors and reraise with a better error message?

@almarklein I'm not sure I understand. What kind of error message would you expect to see here?

Reply:
I mean that the error messages you list above may be rather cryptic and depend a lot on what is provided. If we catch the error where the plugin is instantiated, we can re-raise with something like "the provided plugin does not look like a plugin. Original error message: ..."

Reply:
I think this is already very close to what we are doing:

> If we catch the error where the plugin is instantiated

That's what we do :) the exception handling is a bit lengthy, because we also take care of other causes of failure, e.g. failing because the backend is missing (for real plugins), or plugins rejecting the requires in the indented way (raising `InitializationError`), but it is the same place:

https://github.com/imageio/imageio/blob/924e1c5c43576acf16247c3c47b7baf6c6f68fe6/imageio/core/imopen.py#L148-L164

> "the provided plugin does not look like a plugin. Original error message: ..."

In the current code `generic_error` is the original error message produced by the imposter plugin. Since we `raise from generic_error` its exception will be placed above our exception like in the example above. If that is too verbose, then we can remove that and `raise from None` instead. This would produce the following stack trace:

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\v3_api.py", line 39, in imread
    with imopen(uri, "r", **plugin_kwargs) as img_file:
  File "C:\Users\Sebastian\Documents\Coding-Projects\imageio\imageio\core\imopen.py", line 179, in imopen
    raise err_type(err_msg) from err_from
OSError: An unknown error occured while initializing `<class 'dict'>`
```

We could also place the imposter plugin's stack-trace below our stack trace (just `raise` no `from`), but that seems even more confusing to me.

Regarding the error message itself, we could check for `isinstance(plugin, str)` again and customize the error message to something like `f"The provided plugin does not appear to be a plugin. Got: {plugin}. "`.

Reply:
Aaah! If only I had read 3 more lines downs :)  Sorry about the noise :blush:

Reply:
No worries, I can relate.

Do you think we should change the current exception if the plugin is not a string (i.e, if it is an imposter plugin), or is it fine as is?

Reply:
Yeah, maybe change that error message to mention "pugin", e.g. "Could not initialize plugin`{plugin}`"

Reply:
Done. it is now

```
OSError: An unknown error occured while initializing plugin `<class 'dict'>`.
```